### ISSUE TITLE
Exclude .git directory when uploading template for scaffolder dry-runs

### ISSUE THREAD

**Issue by Pike** [backstage/backstage#12561]
## Expected Behavior

When uploading content for a scaffolder dry-run, I get `payload too large` errors. The easiest way to fix them is to remove the `.git` folder from my local check-out, but that's also complicating things when you work on the templates.

## Actual Behavior

Exclude the `.git` folder from the content uploaded.

## Steps to Reproduce

1. Use the dry-run templating action, uploading a local directory
2. Fill out the form
3. Dry-run the template

## Context

_Originally posted by @leon-vg in https://github.com/backstage/backstage/issues/12176#issuecomment-1168720597_

## Your Environment

backstage 1.3.0

- Browser Information: 

- Output of `yarn backstage-cli info`: 

```text
OS:   Darwin 20.6.0 - darwin/x64
node: v16.15.1
yarn: 1.22.19
cli:  0.17.2 (installed)

Dependencies:
  @backstage/app-defaults                         1.0.3
  @backstage/backend-common                       0.14.0
  @backstage/backend-tasks                        0.3.2
  @backstage/catalog-client                       1.0.3
  @backstage/catalog-model                        1.0.3
  @backstage/cli-common                           0.1.9
  @backstage/cli                                  0.17.2
  @backstage/config-loader                        1.1.2
  @backstage/config                               1.0.1
  @backstage/core-app-api                         1.0.3
  @backstage/core-components                      0.9.5
  @backstage/core-plugin-api                      1.0.3
  @backstage/errors                               1.0.0
  @backstage/integration-react                    1.1.1
  @backstage/integration                          1.2.1
  @backstage/plugin-api-docs                      0.8.6
  @backstage/plugin-app-backend                   0.3.33
  @backstage/plugin-auth-backend                  0.14.1
  @backstage/plugin-auth-node                     0.2.2
  @backstage/plugin-catalog-backend-module-github 0.1.4
  @backstage/plugin-catalog-backend               1.2.0
  @backstage/plugin-catalog-common                1.0.3
  @backstage/plugin-catalog-graph                 0.2.18
  @backstage/plugin-catalog-import                0.8.9
  @backstage/plugin-catalog-react                 1.1.1
  @backstage/plugin-catalog                       1.3.0
  @backstage/plugin-github-actions                0.5.6
  @backstage/plugin-org                           0.5.6
  @backstage/plugin-permission-common             0.6.2
  @backstage/plugin-permission-node               0.6.2
  @backstage/plugin-permission-react              0.4.2
  @backstage/plugin-proxy-backend                 0.2.27
  @backstage/plugin-scaffolder-backend            1.3.0
  @backstage/plugin-scaffolder-common             1.1.1
  @backstage/plugin-scaffolder                    1.3.0
  @backstage/plugin-search-backend-node           0.6.2
  @backstage/plugin-search-backend                0.5.3
  @backstage/plugin-search-common                 0.3.5
  @backstage/plugin-search-react                  0.2.1
  @backstage/plugin-search                        0.9.0
  @backstage/plugin-tech-radar                    0.5.13
  @backstage/plugin-techdocs-backend              1.1.2
  @backstage/plugin-techdocs-node                 1.1.2
  @backstage/plugin-techdocs-react                1.0.1
  @backstage/plugin-techdocs                      1.2.0
  @backstage/plugin-user-settings                 0.4.5
  @backstage/release-manifests                    0.0.4
  @backstage/test-utils                           1.1.1
  @backstage/theme                                0.2.15
  @backstage/types                                1.0.0
  @backstage/version-bridge                       1.0.1

```

**Comment by Pike** [comment]
I'm afraid I'm pretty booked these days, sorry.

**Comment by nicorikken** [comment]
This would be helpful. I tried the editor and it broke my git repo.

**Comment by sabinasiddiqi** [comment]
I am trying to see if I can submit a PR to fix that.

I tried to trace the error and I am see it triggered at these lines: https://github.com/backstage/backstage/blob/master/packages/core-app-api/src/apis/implementations/FetchApi/IdentityAuthInjectorFetchMiddleware.ts#L67
https://github.com/backstage/backstage/blob/master/packages/core-app-api/src/apis/implementations/FetchApi/PluginProtocolResolverFetchMiddleware.ts#L41

I console logged what the value of init.body is at that point and it doesn't include any .git folders as far as I can tell for the template I am using - and we're still getting the payload too large error. Where did you trace that issue from? @Pike 

I am going to try [this solution](https://gist.github.com/Maqsim/857a14a4909607be13d6810540d1b04f) in the [scaffolder-backend router](https://github.com/backstage/backstage/blob/master/plugins/scaffolder-backend/src/service/router.ts#L81) to see if it helps since I am seeing express version > 4 in the yarn.lock files.

**Comment by benjdlambert** [comment]
@sabinasiddiqi are you submitting large files or is it also the fact that your `.git` repo is being included? You should be able to see from the POST body in the browser dev console client side, the request will never make it to the backend which is why the body is empty as the router won't allow it for parsing. I'd really like to avoid changing the limits, as they are there for a reason, and removing them could cause other issues around security.

I think what we probably want to do is filter out a common frequently known set of folders that we should exclude, that being `.git` for a start, but maybe also configurable. I think a reasonable place to start looking would be somewhere around here: https://github.com/backstage/backstage/blob/a2490f64fe18bb718fc461cfc85f2c45b63a9e92/plugins/scaffolder/src/lib/filesystem/WebFileSystemAccess.ts

**Comment by sabinasiddiqi** [comment]
@benjdlambert I tried a really small folder with no .git and the error was still showing. On my local I switched to using the body-parser in the express router and that has seemed to fix the issue.

**Comment by sabinasiddiqi** [comment]
Just to be clear - the fix isn't to change the limits. It is to use the body parser package with the same limits. I was thinking of exporting the const for MAX_CONTENT_SIZE [here](https://github.com/backstage/backstage/pull/12176/files#diff-6928ff0c740df09193a465f32c76049fc80e5002629c4f95b0d39bdf78db5e7dL33) and importing it where the router is used so the same limit is used when setting it.

**Comment by sabinasiddiqi** [comment]
Okay so now that I have been digging into this it looks like there are multiple issues that can lead to this warning:
1) Using the wrong package i.e. for express >-4.0 we should be using the bodyParser in the express router - this one is what my local fix fixes - and before doing that no matter the size of my files, I was getting the same `PayloadTooLargeError: request entity too large` error.

I can have a PR ready for that pretty quickly.

2) The actual size of the template files exceeds the 100kb [express router default limit](https://expressjs.com/en/api.html#express.json) (this is my use case).

@benjdlambert I would need to understand the security/other concerns here because the only fix for my use case I can see is to increase the size limit - the payload size I am dealing with is 128kb.

3) 2 but only because unnecessary folders such as .git are included (the original posters case).

This seems to be the OPs issue and for this we would need to filter out the irrelevant folders as suggested by @benjdlambert 

MAX_CONTENT_SIZE seems to [only apply to one file at a time](https://github.com/backstage/backstage/blob/master/plugins/scaffolder/src/components/TemplateEditorPage/DryRunContext.tsx#L135) and not the complete request body so that is not what is triggering the error I was seeing.

**Comment by Pike** [comment]
I've done some more investigation, and here's what I found:

bodyParser.json checks the limit on the uncompressed stream, and it's using the 100kb default limit from express (express.json is just body-parser.json).

Excluding `.git` is definitely helpful, but just my skeleton is 1M, so the only way I found was to actually specify a bigger limit in https://github.com/backstage/backstage/blob/93fcda09ced3f96b52f462291d8eafd1be72437e/plugins/scaffolder-backend/src/service/router.ts#L151.

I used 10MB, but no idea if that's a genuinely a good idea.

My use-case right now is that I'm hooking up template testing to the dry-run api.
